File: app/src/main/java/com/yc/emotion/home/index/ui/widget/GradientColorTextView.java
package com.yc.emotion.home.index.ui.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Shader;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;
import android.util.TypedValue;

import com.yc.emotion.home.R;


/**
 * Created by suns  on 2019/10/15 19:12.
 */
public class GradientColorTextView extends AppCompatTextView {

    private LinearGradient mLinearGradient;
    private Paint mPaint;
    private int mViewWidth = 0;
    private Rect mTextBound = new Rect();
    private final int startColor;
    private final int endColor;

    public GradientColorTextView (Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray ta= context.obtainStyledAttributes(attrs,R.styleable.GradientColorTextView);

        startColor = ta.getColor(R.styleable.GradientColorTextView_startColor,0xFFFFEABA);
        endColor = ta.getColor(R.styleable.GradientColorTextView_endColor,0xFFBE8B49);

        ta.recycle();

    }

    @Override
    protected void onDraw(Canvas canvas) {
        mViewWidth = getMeasuredWidth();
        mPaint = getPaint();
        String mTipText = getText().toString();
        mPaint.getTextBounds(mTipText, 0, mTipText.length(), mTextBound);
        mLinearGradient = new LinearGradient(0, 0, mViewWidth, 0,
                new int[]{startColor, endColor},
                null, Shader.TileMode.REPEAT);
        mPaint.setShader(mLinearGradient);
        canvas.drawText(mTipText, getMeasuredWidth() / 2 - mTextBound.width() / 2, getMeasuredHeight() / 2 + mTextBound.height() / 2, mPaint);
    }
}

File: app/src/main/java/com/yc/emotion/home/index/ui/widget/GradientColorTextView.kt
package com.yc.emotion.home.index.ui.widget

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import com.yc.emotion.home.R

/**
 * Created by suns  on 2019/10/15 19:12.
 */
class GradientColorTextView(context: Context, attrs: AttributeSet?) : AppCompatTextView(context, attrs) {
    private var mLinearGradient: LinearGradient? = null
    private lateinit var mPaint: Paint
    private var mViewWidth = 0
    private val mTextBound = Rect()
    private val startColor: Int
    private val endColor: Int
    override fun onDraw(canvas: Canvas) {
        mViewWidth = measuredWidth
        mPaint = paint
        val mTipText = text.toString()
        mPaint.getTextBounds(mTipText, 0, mTipText.length, mTextBound)
        mLinearGradient = LinearGradient(0f, 0f, mViewWidth.toFloat(), 0f, intArrayOf(startColor, endColor),
                null, Shader.TileMode.REPEAT)
        mPaint.shader = mLinearGradient
        canvas.drawText(mTipText, measuredWidth / 2 - mTextBound.width() / 2.toFloat(), measuredHeight / 2 + mTextBound.height() / 2.toFloat(), mPaint)
    }

    init {
        val ta = context.obtainStyledAttributes(attrs, R.styleable.GradientColorTextView)
        //        0xFFFFEABA
//        0xFFBE8B49
        startColor = ta.getColor(R.styleable.GradientColorTextView_startColor, 0xFFFFEABA.toInt())
        endColor = ta.getColor(R.styleable.GradientColorTextView_endColor, -0x4174b7)
        ta.recycle()
    }
}